### ISSUE TITLE
role_name check result depends on scan order

### ISSUE THREAD

**Issue by AKorezin** [ansible/ansible-lint#2385]
##### Summary

The result of scanning role from `examples/roles/invalid_due_to_meta` may depend on the order of discovered files.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
ansible --version
ansible [core 2.13.3]
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.10/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.5 (main, Jul 25 2022, 15:52:08) [GCC 11.2.1 20220219]
  jinja version = 3.0.3
  libyaml = True

ansible-lint --version
ansible-lint 6.5.2 using ansible 2.13.3
```

- ansible installation method: OS package
- ansible-lint installation method: OS package

##### OS / ENVIRONMENT

- Archlinux
- custom build alpine:3.16 in docker

##### STEPS TO REPRODUCE

1. Clone ansible-lint repo.
1. Go to `examples/roles/`.
1. Run `ansible-lint -f json -vvv invalid_due_to_meta` several times.
1. Get different results.

##### Desired Behavior

Output should always look like this
```json
[{"type": "issue", "check_name": "role-name", "categories": ["deprecations", "metadata"], "severity": "critical", "description": "Role name invalid-due-to-meta does not match ``^\\[a-z]\\[a-z0-9_]*$`` pattern.", "fingerprint": "578af31a702c8ab20a037b71423cbc38b1a97daa417270b30de6ca67e9ac77db", "location": {"path": "invalid_due_to_meta/meta/main.yml", "lines": {"begin": 1}}}]
```

##### Actual Behavior

Output is not the same, sometimes it will be as desired.
Sometimes it will look like this
```json
[{"type": "issue", "check_name": "role-name", "categories": ["deprecations", "metadata"], "severity": "critical", "description": "Role name invalid-due-to-meta does not match ``^\\[a-z]\\[a-z0-9_]*$`` pattern.", "fingerprint": "78cfbdcc4dd35ad647618a30c7dc2260cd8f5ff9124642cfb9c0dad4298945fa", "location": {"path": "invalid_due_to_meta", "lines": {"begin": 1}}}]
```

This is happened because of different file order.
If the order is
```
DEBUG    Examining invalid_due_to_meta/meta/main.yml of type meta
DEBUG    Examining invalid_due_to_meta of type role
DEBUG    Examining invalid_due_to_meta/tasks/main.yaml of type tasks
```
then the result will be as desired.
If the order is
```
DEBUG    Examining invalid_due_to_meta of type role
DEBUG    Examining invalid_due_to_meta/meta/main.yml of type meta
DEBUG    Examining invalid_due_to_meta/tasks/main.yaml of type tasks
```
then the result will differ.

**Comment by ssbarnea** [comment]
If you rely on output order, use JSON output. There is no guarantee that stdout or stderr will produce identical result as we run lots of tasks in multiple processes.

**Comment by AKorezin** [comment]
Hi @ssbarnea,
Are you sure that this is not json output `ansible-lint -f json -vvv invalid_due_to_meta`?

**Comment by ssbarnea** [comment]
Just remove the -v and use `-q -f json`, that output should be consistent between runs. Still, stderr output is never expected to be consistent between runs and that is what you see with `DEBUG` lines.

What you see on screen is a combination of stdout and stderr, so better to redirect stderr to a file to clean the output. Let me know about the results, I will reopen the bug if needed as I do understand why users might be looking for consistent results. ;)

**Comment by AKorezin** [comment]
This bug was fixed in `6.12`, yet the result is not the best.

Here's my new reproduction script.
```sh
docker run -it --rm python:3.10-slim bash
apt-get update
apt-get install git jq --no-install-recommends -y
cd /tmp
git clone https://github.com/ansible/ansible-lint.git
cd ansible-lint/examples/roles/
pip install 'ansible-lint==6.11.*'
for i in {1..10};do ansible-lint -f json -q invalid_due_to_meta | jq '.[].fingerprint'; echo; done
pip install 'ansible-lint==6.12.*'
for i in {1..10};do ansible-lint -f json -q invalid_due_to_meta | jq '.[].fingerprint'; echo; done
```
The results are
- `6.11.0`
  - ```
    "578af31a702c8ab20a037b71423cbc38b1a97daa417270b30de6ca67e9ac77db"
    "b0353dfeefebe5c100b931c0934a2b3619b772c78a6a720727093e94c62c1df8"
    ```
  - ```
    "78cfbdcc4dd35ad647618a30c7dc2260cd8f5ff9124642cfb9c0dad4298945fa"
    "b0353dfeefebe5c100b931c0934a2b3619b772c78a6a720727093e94c62c1df8"
    ```
- `6.12.2`
  - ```
    "78cfbdcc4dd35ad647618a30c7dc2260cd8f5ff9124642cfb9c0dad4298945fa"
    "578af31a702c8ab20a037b71423cbc38b1a97daa417270b30de6ca67e9ac77db"
    "b0353dfeefebe5c100b931c0934a2b3619b772c78a6a720727093e94c62c1df8"
    ```

Why I'd like to see the result with fingerprint `578af31a`?
Because it's a bit more informative
```json
[
  {
    "type": "issue",
    "check_name": "role-name",
    "categories": [
      "deprecations",
      "metadata"
    ],
    "url": "https://ansible-lint.readthedocs.io/rules/role-name/",
    "severity": "critical",
    "level": "error",
    "description": "Role name invalid-due-to-meta does not match ``^\\[a-z]\\[a-z0-9_]*$`` pattern.",
    "fingerprint": "78cfbdcc4dd35ad647618a30c7dc2260cd8f5ff9124642cfb9c0dad4298945fa",
    "location": {
      "path": "invalid_due_to_meta",
      "lines": {
        "begin": 1
      }
    }
  }
]
```
```json
[
  {
    "type": "issue",
    "check_name": "role-name",
    "categories": [
      "deprecations",
      "metadata"
    ],
    "url": "https://ansible-lint.readthedocs.io/rules/role-name/",
    "severity": "critical",
    "level": "error",
    "description": "Role name invalid-due-to-meta does not match ``^\\[a-z]\\[a-z0-9_]*$`` pattern.",
    "fingerprint": "578af31a702c8ab20a037b71423cbc38b1a97daa417270b30de6ca67e9ac77db",
    "location": {
      "path": "invalid_due_to_meta/meta/main.yml",
      "lines": {
        "begin": 1
      }
    }
  }
]
```

There's also another fingerprint
```json
[
  {
    "type": "issue",
    "check_name": "schema[meta]",
    "categories": [
      "core",
      "experimental"
    ],
    "url": "https://ansible-lint.readthedocs.io/rules/schema/",
    "severity": "blocker",
    "level": "warning",
    "description": "'invalid-due-to-meta' does not match '^\\[a-z]\\[a-z0-9_]+$'",
    "fingerprint": "b0353dfeefebe5c100b931c0934a2b3619b772c78a6a720727093e94c62c1df8",
    "location": {
      "path": "invalid_due_to_meta/meta/main.yml",
      "lines": {
        "begin": 1
      }
    },
    "content": {
      "body": " Returned errors will not include exact line numbers, but they will mention\nthe schema name being used as a tag, like ``schema[playbook]``,\n``schema[tasks]``.\n\nThis rule is not skippable and stops further processing of the file.\n\nIf incorrect schema was picked, you might want to either:\n\n* move the file to standard location, so its file is detected correctly.\n* use ``kinds:`` option in linter config to help it pick correct file type.\n"
    }
  }
]
```
